### ISSUE TITLE
(Regression) Pressing CTRL+C on `docker-compose up` hangs when using containers with restart: unless-stopped|always

### ISSUE THREAD

**Issue by joanbm** [docker/compose#9003]
**Description**

Since version 2.2.0, when I launch a `docker-compose.yml` file defining one or more containers with its restart policy set to `restart: unless-stopped` or `restart: always` with `docker-compose up`, then pressing CTRL+C after the container is up will stop the container as expected, but the docker-compose process will no longer successfully quit. Instead, it will hang, and the only way to get it to quit is to terminate it with something like `pkill -6 docker-compose`.

I have bisected the problem and it appears to have been introduced in commit ed2395819d81ac812a5eefeeccbf77432392626c. Testing a version previous to this commit or reverting it, avoids the problem when terminating with CTRL+C.

**Steps to reproduce the issue:**
1. Install Docker Compose >= 2.2.0  (tested up to latest v2 branch commit)
2. Create a new folder with a `docker-compose.yml` file with the following content:

```yml
version: "3.4"

services:
  db:
    image: alpine:3.15
    command: ["/bin/sleep", "9999"]
    restart: unless-stopped
```

3. Run `docker-compose up` on the folder (where `docker-compose` is the compiled `compose` binary, do *not* run `docker compose up`, as this seems to work around the problem).

4. Wait for the container to start up and then press CTRL+C and wait some seconds.

**Describe the results you received:**

The container stops, but docker-compose does never quit no matter how many times I press CTRL+C. I need to terminate the docker-compose process in order to quit it.

```console
$ docker-compose up
[+] Running 2/2
 ⠿ Network dc_default  Created                                                                                                                                                                                  0.0s
 ⠿ Container dc-db-1   Created                                                                                                                                                                                  0.1s
Attaching to dc-db-1
^CGracefully stopping... (press Ctrl+C again to force)
[+] Running 1/1
 ⠿ Container dc-db-1  Stopped                                                                                                                                                                                  10.4s
[+] Running 0/0
 ⠿ Container dc-db-1  Error while Killing                                                                                                                                                                       0.0s
^C^C^C^C^C^C^C^C^C
```

**Describe the results you expected:**

The container stops, then docker-compose quits gracefully.

```console
$ docker-compose up
[+] Running 2/2
 ⠿ Network dc_default  Created                                                                                                                                                                                  0.0s
 ⠿ Container dc-db-1   Created                                                                                                                                                                                  0.1s
Attaching to dc-db-1
^CGracefully stopping... (press Ctrl+C again to force)
[+] Running 1/1
 ⠿ Container dc-db-1  Stopped                                                                                                                                                                                  10.4s
canceled
```

**Additional information you deem important (e.g. issue happens only occasionally):**

Always reproduces. Reproduced on current Arch Linux and a clean Fedora 35 install. Has begun to happen on commit ed2395819d81ac812a5eefeeccbf77432392626c.

**Output of `docker compose version`:**

```
$ docker compose version
Docker Compose version 2.2.2
```

**Output of `docker info`:**

```
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Build with BuildKit (Docker Inc., v0.6.1-docker)
  compose: Docker Compose (Docker Inc., 2.2.2)

Server:
 Containers: 1
  Running: 0
  Paused: 0
  Stopped: 1
 Images: 4
 Server Version: 20.10.11
 Storage Driver: overlay2
  Backing Filesystem: btrfs
  Supports d_type: true
  Native Overlay Diff: false
  userxattr: true
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1e5ef943eb76627a6d3b6de8cd1ef6537f393a71.m
 runc version: v1.0.2-0-g52b36a2d-dirty
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  rootless
  cgroupns
 Kernel Version: 5.16.0-rc4-12001-mainline-git
 Operating System: Arch Linux (error determining if containerized)
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 31.36GiB
 Name: solpc
 ID: CIRU:4366:6H77:IDAY:WUDP:4AHN:BHDA:7RBP:7HIC:PTNP:FTCY:BFFD
 Docker Root Dir: /home/zealcharm/.local/share/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Default Address Pools:
   Base: 192.168.3.128/25, Size: 28
```

**Additional environment details:**
(Above environment is rootless but I have also reproduced it with rootful Docker.)

**Comment by joanbm** [comment]
PS: Seems related to #8977, not sure if it's actually the same issue...

**Comment by 1ace** [comment]
I'm encountering the same bug, and I have a test-case that can be run from a CI:
```yml
services:

  server:
    image: alpine
    command: /bin/sleep infinity
    restart: always

  client-test:
    image: alpine
    command: /bin/true
```

If you remove the `restart` line:
```
~/bug $ docker compose up --abort-on-container-exit
[+] Running 3/0
 ⠿ Network bug_default          Created 0.0s
 ⠿ Container bug-server-1       Created 0.0s
 ⠿ Container bug-client-test-1  Created 0.0s
Attaching to bug-client-test-1, bug-server-1
bug-client-test-1 exited with code 0
Aborting on container exit...
[+] Running 2/2
 ⠿ Container bug-server-1       Stopped 10.3s
 ⠿ Container bug-client-test-1  Stopped 0.0s
ERRO[0010] 0
~/bug $ 
```

If you leave the `restart` line:
```
~/bug $ docker compose up --abort-on-container-exit
[+] Running 3/0
 ⠿ Network bug_default          Created 0.0s
 ⠿ Container bug-server-1       Created 0.0s
 ⠿ Container bug-client-test-1  Created 0.0s
Attaching to bug-client-test-1, bug-server-1
bug-client-test-1 exited with code 0
Aborting on container exit...
[+] Running 2/2
 ⠿ Container bug-server-1       Stopped 10.2s
 ⠿ Container bug-client-test-1  Stopped 0.0s
ERRO[0010] 0
```
and it never returns.

**Comment by flaviut** [comment]
This also came up here: https://unix.stackexchange.com/questions/680756/docker-compose-up-doesnt-exit-when-containers-stopped

**Comment by verhovensky** [comment]
Same for me, got container restarted after `pkill -6 docker-compose` 

Docker Compose version 2.2.2

`docker version`
Client:
 Version:           20.10.11
 API version:       1.41
 Go version:        go1.17.3
 Git commit:        dea9396e18
 Built:             Sat Nov 20 12:57:52 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.11
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.3
  Git commit:       847da184ad
  Built:            Sat Nov 20 12:57:08 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.5.8
  GitCommit:        1e5ef943eb76627a6d3b6de8cd1ef6537f393a71.m
 runc:
  Version:          1.0.3
  GitCommit:        v1.0.3-0-gf46b6ba2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

`docker info`

Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Build with BuildKit (Docker Inc., v0.6.1-docker)
  compose: Docker Compose (Docker Inc., 2.2.2)

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 4
 Server Version: 20.10.11
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1e5ef943eb76627a6d3b6de8cd1ef6537f393a71.m
 runc version: v1.0.3-0-gf46b6ba2
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.13.19-2-MANJARO
 Operating System: Manjaro Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 5.675GiB
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

**Comment by joanbm** [comment]
According to my testing it's fixed in 740276f550ecb7da444b20dc21860ce2f2c547d6, which is not yet in any released version.

**Comment by MitchTalmadge** [comment]
I'm still running into this problem. Removing all `restart` configurations has fixed it, but that's not ideal.

```
$ docker --version
Docker version 20.10.16, build aa7e414
$ docker compose version
Docker Compose version v2.6.0
```

**Comment by SamD** [comment]
Seeing the same

```
Docker Compose version v2.6.0
```

**Comment by GrayCygnus** [comment]
Joining the choir here, getting this behavior when running `docker-compose up` waiting till everything goes up, and then CTRL+C ... it "gracefully" kills the containers but then hangs and never releases control of the console.

>Docker Compose version v2.6.0

**Comment by AnnoyingTechnology** [comment]
Seeing the same behavior with `Docker Compose version v2.12.2` and `Docker version 24.0.7, build afdd53b`

**Comment by glours** [comment]
Hey @AnnoyingTechnology 👋 
Compose `v2.12.2` is one year old, the latest release is `v2.23.3`, can you try with this one as we have done a lot of fixes in a year.
If you still have the problem, could you open new issue, follow the template so we could have the maximum of information to analyse it please?

**Comment by glours** [comment]
@AnnoyingTechnology glad to know 👍

**Comment by mindplay-dk** [comment]
I'm on `2.24.6` and still having this problem. (I'm on Ubuntu 22.04 LTS on WSL2 if that matters.)

**Comment by kchawla-pi** [comment]
I have recently started facing this problem on Windows. Only solution is to close the Terminal window.
Docker Compose version v2.24.5-desktop.1

**Comment by RobertSkawinski** [comment]
same here for example with mailcow

first `docker compose up`
then everything works great and after pressing `<CTRL> + <C> `
even when forcing to kill nothing changes - only closing the terminal window helps
```
^CGracefully stopping... (press Ctrl+C again to force)
Aborting on container exit...
[+] Stopping 19/19
 ✔ Container mailcowdockerized-olefy-mailcow-1      Stopped                                                                          10.2s 
 ✔ Container mailcowdockerized-ofelia-mailcow-1     Stopped                                                                           0.3s 
 ✔ Container mailcowdockerized-ipv6nat-mailcow-1    Stopped                                                                           0.0s 
 ✔ Container mailcowdockerized-dockerapi-mailcow-1  Stopped                                                                           0.7s 
 ✔ Container mailcowdockerized-watchdog-mailcow-1   Stopped                                                                           0.0s 
 ✔ Container mailcowdockerized-solr-mailcow-1       Stopped                                                                           0.7s 
 ✔ Container mailcowdockerized-rspamd-mailcow-1     Stopped                                                                           0.4s 
 ✔ Container mailcowdockerized-clamd-mailcow-1      Stopped                                                                           0.0s 
 ✔ Container mailcowdockerized-memcached-mailcow-1  Stopped                                                                           1.0s 
 ✔ Container mailcowdockerized-acme-mailcow-1       Stopped                                                                           0.0s 
 ✔ Container mailcowdockerized-postfix-mailcow-1    Stopped                                                                           0.0s 
 ✔ Container mailcowdockerized-nginx-mailcow-1      Stopped                                                                           0.4s 
 ✔ Container mailcowdockerized-php-fpm-mailcow-1    Stopped                                                                           0.3s 
 ✔ Container mailcowdockerized-sogo-mailcow-1       Stopped                                                                          10.2s 
 ✔ Container mailcowdockerized-dovecot-mailcow-1    Stopped                                                                          10.5s 
 ✔ Container mailcowdockerized-redis-mailcow-1      Stopped                                                                           0.3s 
 ✔ Container mailcowdockerized-mysql-mailcow-1      Stopped                                                                           0.4s 
 ✔ Container mailcowdockerized-unbound-mailcow-1    Stopped                                                                           0.1s 
 ✔ Container mailcowdockerized-netfilter-mailcow-1  Stopped                                                                           4.2s 


^[
^Cno container to kill^C^C^C^C
```
Here some System details:
```
# docker compose version
Docker Compose version v2.18.0

# docker version
Client: Docker Engine - Community
 Version:           26.0.0
 API version:       1.45
 Go version:        go1.21.8
 Git commit:        2ae903e
 Built:             Wed Mar 20 15:17:48 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          26.0.0
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.8
  Git commit:       8b79278
  Built:            Wed Mar 20 15:17:48 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

# cat /etc/os-release 
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

**Comment by NEGU93** [comment]
> According to my testing it's fixed in [740276f](https://github.com/docker/compose/commit/740276f550ecb7da444b20dc21860ce2f2c547d6), which is not yet in any released version.

Is this an issue then or not? If so, from which version should it work? If the fix was for Docker Compose version lower than v2.18.0 this should be re-opened.

**Comment by joanbm** [comment]
@NEGU93 I believe it should have been released in Docker Compose v2.2.3 in early 2022. So, a long time ago.

The recent influx of comments in the last few months suggest to me that the issue was likely solved during 2022-2023 and recently some similar problem recurred but it may have a different root cause. So, I suggest that you open a new issue if there isn't one already.

**Comment by alexanderzaiets-paysera** [comment]
same issue with compose version 2.4.1
